Support string literals with c-style character espaces

smakalias opened the pull request:
## Summary of the changes / Why this improves CrateDB

Support string literals with c-style character escapes

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

seut commented on sql-parser/src/main/java/io/crate/sql/Literals.java:
```diff
@@ -24,8 +24,17 @@
+    private static Pattern octalPattern = Pattern.compile("\\\\([0-7]{1,3})");
+    private static Pattern hexPattern = Pattern.compile("\\\\x([0-9A-F]{1,2})");
+    private static Pattern uni16Pattern = Pattern.compile("\\\\u([0-9A-F]{4})");
+    private static Pattern uni32Pattern = Pattern.compile("\\\\U([0-9A-F]{8})");
```
Afaik the case of `U` does not define if 16 or 32bit, it could be used for both.

mfussenegger commented on sql-parser/src/main/java/io/crate/sql/Literals.java:
```diff
@@ -34,4 +43,47 @@ public static String quoteStringLiteral(String literal) {
+        if (!more) {
+            return input;
+        }
+        StringBuffer stringBuffer = new StringBuffer(input.length());
```
For single threaded code it would be better to use `StringBuilder`, as it doesn't have synchronization overhead.

The `replaceEscapedChars` logic looks generally a bit expensive as it does multiple pattern matches. I wonder if it would be justified to do a manual loop over the characters and look for the sequences, so we have only one pass?

smakalias replied:
> The `replaceEscapedChars` logic looks generally a bit expensive as it does multiple pattern matches. I wonder if it would be justified to do a manual loop over the characters and look for the sequences, so we have only one pass?

It's something I wanted to discuss further as I also saw it, initially I decided to go for a solution with less complexity. Do you anticipate this to be used heavily?

smakalias replied:
> For single threaded code it would be better to use `StringBuilder`, as it doesn't have synchronization overhead.

in java 1.8 only the StringBuffer variants are supported in Mather class. Should we break bwd compatibility?

mfussenegger replied:
I think this should still go into 3.3, so no, we shouldn't break compatibility.

I don't think this will be used heavily, but I'm not sure if a more imperative/direct approach is really more complex.

I'd suggest to give it a go and then we can see how it compares.

mfussenegger commented on sql-parser/src/main/java/io/crate/sql/tree/EscapedCharStringLiteral.java:
```diff
@@ -0,0 +1,43 @@
+
+import io.crate.sql.Literals;
+
+public class EscapedCharStringLiteral extends StringLiteral {
```
Do we need this inheritance?

smakalias replied:
to me, it seemed a natural extension as we already have the `value` and just wanted in addition to keep the `raw_value`. I also thought that an `EscapedStringLiteral` is a `StringLiteral`

smakalias commented on sql-parser/src/main/antlr/SqlBase.g4:
```diff
@@ -303,6 +304,10 @@ nullLiteral
     ;
 
+escapedCharsStringLiteral
+    : 'E' STRING
```
yes, it does - I will add a test to make it clear

smakalias commented:
@seut please have another look, compared to the previous commit, I removed the `String::split` and mainly worked with character arrays instead of strings to avoid copying.

smakalias commented on blackbox/docs/sql/general/lexical-structure.rst:
```diff
@@ -58,25 +92,26 @@ quoted if used as identifiers.
+    INDEX, INNER, INPUT, INSERT
+    INT, INTEGER, INTERSECT, INTO
+    IP, IS, JOIN, LAST
+    LEFT, LICENSE, LIKE, LIMIT
```
I also added the `LICENSE` here (it was missing)

smakalias commented on sql-parser/src/main/java/io/crate/sql/Literals.java:
```diff
@@ -24,14 +24,163 @@
+            while (current < inputCharArray.length && '\\' != inputCharArray[current]) {
+                current++;
+            }
+            tokenCharArray = Arrays.copyOfRange(inputCharArray, prev, current);
```
How can I remove this copy operation here? I just need a reference for this sub-array.

seut replied:
you can use `builder.append(inputCharArray, prev, current)` instead of creating a new array slice.

smakalias commented on sql-parser/src/main/java/io/crate/sql/Literals.java:
```diff
@@ -24,14 +24,163 @@
+        return idx;
+    }
+
+    private static int parseInt(char[] s, int beginIndex, int endIndex, int radix) throws NumberFormatException {
```
added this shorten version of `Integer::parseInt` as I could not find one in standard lib operating on `char[]`. Is there an alternative that you know of?

mfussenegger commented on sql/src/test/java/io/crate/integrationtests/SelectEscapedCharactersITest.java:
```diff
@@ -0,0 +1,109 @@
+import static org.hamcrest.Matchers.is;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 1)
+public class SelectEscapedCharactersITest extends SQLTransportIntegrationTest {
```
What exactly are we testing here with these integration tests? Seems that they're either redundant or incomplete (and on a too high level = very expensive).

If we want to test the indexing and querybility we can create a unit test using the `QueryTester`, but that would amount to testing Lucene and I'm not sure if that is what we should be doing?

smakalias replied:
@mfussenegger yes, I was curious to test the querybiliy and make sure it works as expected, that was the reason for this integration test. I do agree though that it's too expensive and perhaps a bit redundant to keep thus I removed it. 

Did not know about `QueryTester` so thanks for that. 
Do you feel that more tests are required?

mfussenegger commented on sql-parser/src/main/java/io/crate/sql/Literals.java:
```diff
@@ -24,14 +24,191 @@
+        boolean firstToken = true;
+        int nextTokenIdx, tokenBeginIdx, tokenLen;
+
+        while (current < inputCharArray.length) {
```
I'm probably missing something here, but wouldn't it be possible to go through the characters one by one like this:

```
        StringBuilder result = new StringBuilder();
        int length = input.length();
        for (int i = 0; i < length; i++) {
            char currentChar = input.charAt(i);
            if (currentChar == '\\' && i + 1 < length) {
                char nextChar = input.charAt(i + 1);
                case (nextChar) {
                ....
```

And if there is a escaped sequence do something along the lines of

```
                    int codePoint = Integer.parseInt(x, i + 2, i + 2 + numCharsInSeq, 16);
                    result.append(Character.toChars(codePoint));
                    i += 1 + numCharsInSeq; // skip the extra chars consumed
```

This avoids most copies, doesn't require a custom `parseInt` and should be fairly straight forward?

smakalias replied:
thanks for the input. It's true, no reason to further keep the tokenization logic, what you proposed makes much more sense (and we do get rid of the custom `parseInt` !). +1

mfussenegger commented on sql-parser/src/main/java/io/crate/sql/Literals.java:
```diff
@@ -24,14 +24,140 @@
+        }
+
+        StringBuilder builder = new StringBuilder(length);
+        int endIdx;
```
Could probably reduce the scope of `endIdx` to the places where it is used, to make it clear that it isn't used outside of the specific case blocks.

smakalias replied:
I would rather not change this: `endIdx` is used in 3 main use cases. 
In addition, this would require to either use different name in each case block, or use anonymous blocks or declare it only in the first case where it's used (which I personally find more confusing).
In addition, it's in a for loop, so I'd rather declare it once.

mfussenegger commented on sql-parser/src/test/java/io/crate/sql/LiteralsTest.java:
```diff
@@ -47,4 +47,87 @@ public void testQuote() throws Exception {
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testInvalidHexEscapedUnicode32ThrowsException() throws Exception {
+        Literals.replaceEscapedChars("\\U0000006G");
```
Can we also have a `expectedException.expectMessage` here in these tests, to control the error message?

mfussenegger commented on sql-parser/src/test/java/io/crate/sql/LiteralsTest.java:
```diff
@@ -47,4 +47,87 @@ public void testQuote() throws Exception {
+    }
+
+    @Test
+    public void testReplaceEscapedCharsHexByteValues() throws Exception {
```
Personally I'd prefer it if we could split this up into multiple tests with method names that are more specific about what aspect is being tested. A good question to ask yourself is if based on the test method name you'd be able to re-create the test method body and implement the logic that is being tested.

mfussenegger commented on sql-parser/src/main/java/io/crate/sql/Literals.java:
```diff
@@ -24,14 +24,141 @@
+    }
+
+    /**
+     * Consumes a maximum number of characters of a char sequence matching a given predicate
```
Thought I had commented on this already, but I guess I didn't save the comment.

It's good to see some comments, but I think in this case the comments are only repeating what the signature already states.

Much more interesting would be *why* it is okay to skip some characters. Maybe include some examples.
The method name is probably also a bit miss-leading. It is not consuming any characters (consume implies the things are being eaten up --> gone afterwards). But it calculates the endIdx / amount of characters that are part of the escape sequence.

mfussenegger commented on blackbox/docs/sql/general/lexical-structure.rst:
```diff
@@ -39,6 +39,40 @@ single quotes, e.g. ``'Jack''s car'``.
+literal with the letter ``E`` or ``e``, for example, ``e'hello\nWorld'``.
+The following escaped sequences are supported:
+
+==================================================   ================
```
Is this formatted correctly as table in sphinx ?

smakalias replied:
Yes, I used the [simple table](http://docutils.sourceforge.net/docs/user/rst/quickref.html#tables) syntax

see 
<img width="926" alt="image" src="https://user-images.githubusercontent.com/41899848/53347016-e974e680-3920-11e9-8b7f-9ffb55b68e05.png">
